sdcm/kafka/kafka_producer.py

```diff
@@ -0,0 +1,410 @@
+        # producer_config['batch.num.messages'] = int(
+        #     producer_properties.get('batch.num.messages', kwargs.get('batch_num_messages', 10000)))
+        # producer_config['acks'] = producer_properties.get('acks', kwargs.get('acks', '1'))
+        # producer_config['compression.type'] = producer_properties.get(
+        #     'compression.type', kwargs.get('compression_type', 'none'))
+
+        producer_config.update(producer_properties)
+
+        self.producer = Producer(producer_config)
+        self.wait_for_topic(self.topic, timeout=60)
+
+    def get_topics(self) -> list[str]:
+        admin_client = Producer({'bootstrap.servers': ','.join(self.kafka_addresses)})
+        topics = admin_client.list_topics(timeout=10).topics.keys()
+        LOGGER.debug("Available topics: %s", topics)
+        return topics
+
+    def wait_for_topic(self, topic: str, timeout: int) -> None:
+        wait_for(lambda: topic in self.get_topics(), text=f"waiting for topic={topic}", timeout=timeout)
+
+    def delivery_report(self, err: str, record: Message) -> None:
+        if err:
+            self.failed_count += 1
+            LOGGER.error("Kafka record delivery failed: %s", err)
+        else:
+            self.produced_count += 1
+            LOGGER.debug(
+                "Kafka record delivered to %s [%s] at offset %s", record.topic(), record.partition(), record.offset())
+
+    def run(self) -> None:
+        start_time = time.time()
+        while not self.termination_event.is_set():
+            if self.num_records and self.produced_count >= self.num_records:
+                LOGGER.info("Requested number of records is produced, stopping Kafka producer thread.")
+                self.stop()
+                break
+
+            if self.record_format == 'avro':
+                key = self.generate_avro_key(self.current_key)
+                value = self.generate_avro_value(self.current_key)
+                key_serialized = (
+                    self.key_serializer(
+                        key, SerializationContext(self.topic, MessageField.KEY)
+                    ) if self.key_serializer else None)
+                value_serialized = (
+                    self.value_serializer(
+                        value, SerializationContext(self.topic, MessageField.VALUE)
+                    ) if self.value_serializer else None)
+            elif self.record_format == 'json-with-schema':
+                if self.key_schema:
+                    key = self.generate_json_with_schema_key(self.current_key)
+                    key_serialized = self.key_serializer(key, SerializationContext(self.topic, MessageField.KEY))
+                else:
+                    key_serialized = None
+                value = self.generate_json_with_schema_value(self.current_key)
+                value_serialized = self.value_serializer(value, SerializationContext(self.topic, MessageField.VALUE))
```

Comment:
this is very inefficient. Because `record_format` does not change, we could upfront create key and value generators once and use it in this loop.
Note, that each time we pass this loop, also `generate_avro_key` and `generate_avro_value` loads the same json.

Reply:
Fixed. Deserialized the schemas once in constructor.

Reply:
Good. See you can go even further by crating `key_generator` and `value_generator` before while loop starts to further optimize. But again, if this is only for functional test then this implementation is fine.